### ISSUE TITLE
Finetuning llama2 works but llama2-chat-hf doesn't? Loss goes to zero

### ISSUE THREAD

**Issue by iamshnoo** [huggingface/trl#834]
Library versions:
```
Transformers version: 4.31.0
Datasets : 2.13.0
PEFT version: 0.4.0
TRL version : 0.7.1
PyTorch version : 2.0.1
Accelerate : 0.21.0
Bits and Bytes : 0.40.2
```

Problem: 

I used a combination of the above libraries and the steps outlined in [this blog by Philipp Schmidd](https://www.philschmid.de/instruction-tune-llama-2) to first instruction fine-tune the llama2 models (not the chat versions) on alpaca data. This works great. And I get decent training curves.

<img width="485" alt="Screenshot 2023-10-05 at 3 24 15 AM" src="https://github.com/huggingface/trl/assets/45921510/dd9a4822-9634-4bce-b353-ca902c18b816">

Next, I wanted to use the same script to instruction fine-tune the chat models on alpaca data. For this, the only change should be the name of the model and the way the prompt is formatted (by including the [INST], [\INST], etc tokens) and the rest of the script should work exactly the same way? 

Idk why but it doesn't. Here's the training curve I get.

<img width="493" alt="Screenshot 2023-10-05 at 3 28 58 AM" src="https://github.com/huggingface/trl/assets/45921510/1a2780b2-4aa4-4e2d-a793-1347610c6c5b">

And here is a corresponding minimal code snippet to illustrate my setup

<details>
  <summary>Click to view code</summary>
 
  ```python
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
from peft import LoraConfig, prepare_model_for_kbit_training, get_peft_model
from transformers import TrainingArguments
from trl import SFTTrainer, DataCollatorForCompletionOnlyLM
from datasets import load_dataset
import random
import numpy as np

SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
torch.cuda.manual_seed_all(SEED)

model_id = "meta-llama/Llama-2-7b-chat-hf"

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_use_double_quant=True,
    bnb_4bit_compute_dtype=torch.bfloat16,
)

model = AutoModelForCausalLM.from_pretrained(
    model_id,
    quantization_config=bnb_config,
    use_cache=False,
    device_map="auto",
)

model.config.pretraining_tp = 1

tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"

peft_config = LoraConfig(
    lora_alpha=16,
    lora_dropout=0.1,
    r=64,
    bias="none",
    task_type="CAUSAL_LM",
)

model = prepare_model_for_kbit_training(model)
model = get_peft_model(model, peft_config)

args = TrainingArguments(
    num_train_epochs=1,
    per_device_train_batch_size=4,
    gradient_accumulation_steps=2,
    gradient_checkpointing=True,
    optim="paged_adamw_32bit",
    logging_steps=10,
    save_strategy="epoch",
    learning_rate=3e-4,
    bf16=True,
    tf32=True,
    max_grad_norm=0.3,
    warmup_ratio=0.03,
    lr_scheduler_type="constant",
    disable_tqdm=True,
)

def format_prompt(instruction, sys_prompt=None, input=None, output=None):
    if sys_prompt is None:
        DEFAULT_SYSTEM_PROMPT = """\
            You are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe.  Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.\n\nIf a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\
            """
    else:
        DEFAULT_SYSTEM_PROMPT = sys_prompt
    PROMPT_DICT = {
        "prompt_input": (
            "<s>[INST] <<SYS>>\n{system_prompt}\n<</SYS>>\n"
            "Below is an instruction that describes a task, paired with an input that provides further context. "
            "Write a response that appropriately completes the request.\n\n"
            "### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:\n[/INST]\n{output}"
        ),
        "prompt_no_input": (
            "<s>[INST] <<SYS>>\n{system_prompt}\n<</SYS>>\n"
            "Below is an instruction that describes a task. "
            "Write a response that appropriately completes the request.\n\n"
            "### Instruction:\n{instruction}\n\n### Response:\n[/INST]\n{output}"
        ),
    }


    if input is None or input == "":
        return PROMPT_DICT["prompt_no_input"].format_map(
            {"instruction": instruction, "system_prompt": DEFAULT_SYSTEM_PROMPT, "output": output}
        )
    else:
        return PROMPT_DICT["prompt_input"].format_map(
            {"instruction": instruction, "input": input, "system_prompt": DEFAULT_SYSTEM_PROMPT, "output": output}
        )

max_seq_length = 2048

data_path = "yahma/alpaca-cleaned"
dataset = load_dataset(data_path)
dataset = dataset["train"]

dataset = dataset.map(
    lambda example: {
        "text": format_prompt(
            example["instruction"], example["input"], example["output"]
        ),
    },
    remove_columns=["instruction", "input", "output"],
)

splits = dataset.train_test_split(test_size=0.2)
train_dataset = splits["train"]
eval_dataset = splits["test"]

data_collator = DataCollatorForCompletionOnlyLM(
    response_template=tokenizer("[/INST]")['input_ids'][2:],
    tokenizer=tokenizer,
)

trainer = SFTTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    dataset_text_field="text",
    data_collator=data_collator,
    peft_config=peft_config,
    max_seq_length=max_seq_length,
    tokenizer=tokenizer,
    args=args,
)

trainer.train()
  ```
</details>

I also tried adding a special PAD token that is not EOS by doing 
```
tokenizer.add_special_tokens({'pad_token': '[PAD]'})
model.resize_token_embeddings(len(tokenizer))
```
but I get the same issue even then.

Maybe I am missing something here. But it would be very helpful if someone  (Maybe @younesbelkada, @lewtun @lvwerra ?) could help me figure this out! Thanks.

**Comment by lvwerra** [comment]
What if you remove the template just for the sake to exclude it as a source of issues?

**Comment by iamshnoo** [comment]
The issue persists even if I remove the template (INST, SYS, etc) and replace it with just the basic alpaca formatting.

**Comment by iamshnoo** [comment]
@OneCodeToRuleThemAll Thanks for the pointers. But if you see the code snippet I included in the issue description, I am already doing 
```
tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"
```

and still have the loss going to zero. I think the rest of the script I have is similar to the one you linked to. Am I missing something else?

The same problem persists for me if I do
```
tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = "[PAD]"
tokenizer.padding_side = "right"
model.resize_token_embeddings(len(tokenizer))
```

**Comment by iamshnoo** [comment]
I figured out the problem. It was a mistake on my end.

In the format_prompt function I was using for alpaca, I specified the order of parameters incorrectly, such that the model was only getting the instructions from the dataset, but not the corresponding outputs. 

I tried using my script now on another dataset "timdettmers/openassistant-guanaco" as well and it works for that data as well.
